In [33]:
#здесь считываются исходные данные. Корректировать этот фрагмент кода не следует.
import re
import requests
# import pandas as pd
# import numpy as np


#ссылка на необходимый документ
task_url = 'https://www.gutenberg.org/cache/epub/16527/pg16527.txt'
#отправляем запрос и получаем весь текст документа
raw = requests.get(task_url).text

#индекс начала задач в тексте
#start_index = re.search(r'Все задачи снабжены ответами — сразу после текста задачи.', raw).end()
#индекс конца задач в тексте
#end_index = re.search(r'End of the Project Gutenberg EBook of 1001', raw).start()

In [34]:
start_index = re.search(r'Все задачи снабжены ответами — сразу после текста задачи.', raw).end()
end_index = re.search(r'END OF THE PROJECT GUTENBERG EBOOK 1001 ЗАДАЧА ДЛЯ УМСТВЕННОГО СЧЕТА', raw).start()

In [35]:
#имеем строку, содержащую тексты всех задач
task = raw[start_index: end_index]
task

"\r\n\r\n1. В течение февраля я прочел весь Ветхий Завет. Прочитывал я по 36\r\nстраниц в день. Сколько страниц в Ветхом Завете? [1008]\r\n\r\n2. Я купил 11 десятин земли по 23 руб. и 13 десятин по 19 руб. Сколько я\r\nзаплатил? [500 руб.]\r\n\r\n3. Я купил 18 десятин земли по 18 руб. и 12 десятин по 23 руб. Сколько я\r\nзаплатил? [600 руб.]\r\n\r\n4. Я купил 13 десятин земли по 39 руб. и 17 десятин по 29 руб. Сколько я\r\nзаплатил? [1000 руб.]\r\n\r\n5. Куплено 16 десятин леса по 97 руб. и 26 десятин пахоты по 48 руб.\r\nСколько стоит вся земля? [2800 руб.]\r\n\r\n6. Куплено 31 десятина леса по 32 руб. и 28 десятин по 36 руб. Сколько\r\nстоит вся земля? [2000 руб.]\r\n\r\n7. Я купил 37 десятин земли по 37руб. и 11 десятин по 21 руб. Сколько я\r\nзаплатил? [1600 руб.]\r\n\r\n8. Некто был учителем в течение 14 лет. Сколько дней он учительствовал?\r\n[5110]\r\n\r\n9. Я купил 32 десятины земли, и каждая обошлась мне в 31 руб. 25 коп.\r\nСколько я заплатил? [1000 руб.]\r\n\r\n10. Сколько в

Задание 1. Создайте список, в каждом элементе которого будет храниться ровно по одной задаче.

Иными словами, необходимо исходныю строку `task` разбить таким образом, чтобы каждый элемент разбиения содержал ровно по одной задаче

In [36]:
# изучаем содержимое строки и обращаем внимание,
# что перед началом каждом задачи идет одна и также последовательность символов `\r\n\r\n`
# воспользуемся методом split() и разделим строку по этой последовательности
lst_task = task.split('\r\n\r\n')
len(lst_task)

1006

In [37]:
#на выходе получили список из 1006 задач! Много! Значит где-то есть ошибка
#посмотрим на получившийся список
#обращаем внимание, что в начале и конце списка, есть пустые задачи.
#Давайте их уберем
print(lst_task)

['', '1. В течение февраля я прочел весь Ветхий Завет. Прочитывал я по 36\r\nстраниц в день. Сколько страниц в Ветхом Завете? [1008]', '2. Я купил 11 десятин земли по 23 руб. и 13 десятин по 19 руб. Сколько я\r\nзаплатил? [500 руб.]', '3. Я купил 18 десятин земли по 18 руб. и 12 десятин по 23 руб. Сколько я\r\nзаплатил? [600 руб.]', '4. Я купил 13 десятин земли по 39 руб. и 17 десятин по 29 руб. Сколько я\r\nзаплатил? [1000 руб.]', '5. Куплено 16 десятин леса по 97 руб. и 26 десятин пахоты по 48 руб.\r\nСколько стоит вся земля? [2800 руб.]', '6. Куплено 31 десятина леса по 32 руб. и 28 десятин по 36 руб. Сколько\r\nстоит вся земля? [2000 руб.]', '7. Я купил 37 десятин земли по 37руб. и 11 десятин по 21 руб. Сколько я\r\nзаплатил? [1600 руб.]', '8. Некто был учителем в течение 14 лет. Сколько дней он учительствовал?\r\n[5110]', '9. Я купил 32 десятины земли, и каждая обошлась мне в 31 руб. 25 коп.\r\nСколько я заплатил? [1000 руб.]', '10. Сколько вершков в 375 аршинах? [6000]', '11. Я з

In [38]:
#Убираем пустые задачи вида ''
for t in lst_task: # перебираем последовательно список задач
    if len(t.strip()) == 0:
       lst_task.remove(t) #то мы нашли пустую задачу (строку), удаляем ее из списка

In [39]:
len(lst_task)
#имеем уже 1003 задачи, а это уже близко к 1001.
#Посмотрим на конец и начало списка снова.
# Обращаем внимание, что в конце есть ....
# Думаем как удалить
lst_task

['1. В течение февраля я прочел весь Ветхий Завет. Прочитывал я по 36\r\nстраниц в день. Сколько страниц в Ветхом Завете? [1008]',
 '2. Я купил 11 десятин земли по 23 руб. и 13 десятин по 19 руб. Сколько я\r\nзаплатил? [500 руб.]',
 '3. Я купил 18 десятин земли по 18 руб. и 12 десятин по 23 руб. Сколько я\r\nзаплатил? [600 руб.]',
 '4. Я купил 13 десятин земли по 39 руб. и 17 десятин по 29 руб. Сколько я\r\nзаплатил? [1000 руб.]',
 '5. Куплено 16 десятин леса по 97 руб. и 26 десятин пахоты по 48 руб.\r\nСколько стоит вся земля? [2800 руб.]',
 '6. Куплено 31 десятина леса по 32 руб. и 28 десятин по 36 руб. Сколько\r\nстоит вся земля? [2000 руб.]',
 '7. Я купил 37 десятин земли по 37руб. и 11 десятин по 21 руб. Сколько я\r\nзаплатил? [1600 руб.]',
 '8. Некто был учителем в течение 14 лет. Сколько дней он учительствовал?\r\n[5110]',
 '9. Я купил 32 десятины земли, и каждая обошлась мне в 31 руб. 25 коп.\r\nСколько я заплатил? [1000 руб.]',
 '10. Сколько вершков в 375 аршинах? [6000]',
 '1

In [40]:
#сделаем вывод одно задачи по индексу
lst_task[0]
# Обратим внимание на шаблон задачи.
# 'Номер задачи. Текст задачи? [ответ]'
# Выдвигаем предположение, что текст задачи всегда начинается с номера.
# Следовательно, тот элемент списка задача, в ктором на первом месте идет число.
# Попробуем воспользоваться нашей догадкой

#Однако прежде, как взять число. Замечаем что после числа идет точка.
#Уже знакомый метод split по точке
lst_task[0].split('.')
#Дает разбиение и первый элемент это число, выраженное строкой.
lst_task[0].split('.')[0]
#воспользуемся теперь этой информацией

'1'

In [41]:
new_lst_task = list()
for t in lst_task: # перебираем последовательно список задач
    if t.split('.')[0].isdigit(): #пользуемся методом isdigit() -- проверяет, является ли строка числом
        new_lst_task.append(t) #добавляем в новый список

In [42]:
#проверяем длину. Все!
len(new_lst_task)

999

In [ ]:
#ну или более короткое решение
r = [t for t in task.split('\r\n\r\n') if (len(t) > 0) & (t.split('.')[0].isdigit())]

In [ ]:
r = list()
for t in task.split('\r\n\r\n'):
    if (len(t) > 0) & (t.split('.')[0].isdigit()):
        r.append(t)

['1. В течение февраля я прочел весь Ветхий Завет. Прочитывал я по 36\r\nстраниц в день. Сколько страниц в Ветхом Завете? [1008]',
 '2. Я купил 11 десятин земли по 23 руб. и 13 десятин по 19 руб. Сколько я\r\nзаплатил? [500 руб.]',
 '3. Я купил 18 десятин земли по 18 руб. и 12 десятин по 23 руб. Сколько я\r\nзаплатил? [600 руб.]',
 '4. Я купил 13 десятин земли по 39 руб. и 17 десятин по 29 руб. Сколько я\r\nзаплатил? [1000 руб.]',
 '5. Куплено 16 десятин леса по 97 руб. и 26 десятин пахоты по 48 руб.\r\nСколько стоит вся земля? [2800 руб.]',
 '6. Куплено 31 десятина леса по 32 руб. и 28 десятин по 36 руб. Сколько\r\nстоит вся земля? [2000 руб.]',
 '7. Я купил 37 десятин земли по 37руб. и 11 десятин по 21 руб. Сколько я\r\nзаплатил? [1600 руб.]',
 '8. Некто был учителем в течение 14 лет. Сколько дней он учительствовал?\r\n[5110]',
 '9. Я купил 32 десятины земли, и каждая обошлась мне в 31 руб. 25 коп.\r\nСколько я заплатил? [1000 руб.]',
 '10. Сколько вершков в 375 аршинах? [6000]',
 '1

Задание 2. Согласно названию книги в ней содержится 1001 задача. На самом деле, если Вы все сделали верно, в результирующем списке оказалось 999 задач. Укажите номера отсутствующих задач.

Задание 3. Проверьте есть ли повторяющиеся номера задач. Если таковые имеются, то укажите их?

Задание 4. Проверьте есть ли повторяющиеся текст задач. Если таковые имеются, то выведите их?

#Разбор кода

Задание 2. Напишите функцию для нахождения суммы `n` элементов списка.  Если `n` не задано, то функция должна возвращать сумму всех элементов списка. Список значений и `n` задаются пользователем произвольно (либо через клавиатуру, либо через переменные).

In [ ]:
#написанный код
def s(a:list, n=None):
    # x=a.copy()
    # a.sort()
    #y=a[:n]
    # sum2=0
    # for z in a[:n]:
    #     sum2 += z
    return f'Сумма элементов {a} равна {sum(a[:n])}'

In [ ]:
#написанный код
t1 = 0
example_1 = 10
summ = 0
while t1 < example_1:
    if t1 % 2 == 0:
        summ = summ + t1
        #lst_odd.append(t1)
        t1 += 1